ftw/solr/upgrades/20200401111512_update_modified_index/upgrade.py

```diff
@@ -0,0 +1,159 @@
+        value = value[:-5] + 'Z'
+    return value
+
+
+def solr_date_old(date):
+    """Solr date representation. Fractional seconds are stripped if 0."""
+    value = to_iso8601_old(date)
+    if value.endswith('.000Z'):
+        value = value[:-5] + 'Z'
+    return value
+
+
+commit_interval = 100
+
+
+class UpdateModifiedIndex(UpgradeStep):
+    """Update modified index.
+    """
+
+    def __call__(self):
+        self.logger = logging.getLogger('ftw.upgrade')
+
+        if not self.is_enabled():
+            self.log("Skipping: Solr is not enabled.")
+            return
+        self.sync()
+
+    def is_enabled(self):
+        registry = queryUtility(IRegistry)
+        settings = registry.forInterface(ISolrSettings)
+        return settings.enabled
+
+    def diff(self, solr_date):
+        """Diff with portal catalog"""
+
+        catalog = getToolByName(self.portal, 'portal_catalog')
+        items = catalog.unrestrictedSearchResults()
+        catalog_modified = set(
+            [(item.UID, solr_date(item.modified)) for item in items])
+
+        conn = self.manager.connection
+        resp = conn.search({
+            u'query': u'*:*',
+            u'limit': 10000000,
+            u'params': {u'fl': ['UID', 'modified']},
+        })
+        solr_modified = set(
+            [(doc['UID'], doc.get('modified', u'2000-01-01T00:00:00.000Z'))
+             for doc in resp.docs])
+
+        not_in_sync = set((item[0] for item in catalog_modified - solr_modified))
+        return not_in_sync
+
+    def sync(self, commit_interval=100, idxs=None, doom=True):
+        """Sync Solr with portal catalog"""
+        if not self.is_enabled():
+            return 'Solr indexing is disabled.'
+
+        catalog = getToolByName(self.portal, 'portal_catalog')
+        not_in_sync_old = self.diff(solr_date_old)
+        not_in_sync = self.diff(solr_date)
+
+        # We only want to reindex modified for objects that would
+        # be in sync if we hadn't modified the rounding in solr_date
+        to_correct = not_in_sync - not_in_sync_old
+
+        # Reindex modified for Solr items that should be in sync
+        processed = 0
+        real = timer()
+        lap = timer()
+        cpu = timer(clock)
+
+        if doom:
+            transaction.doom()
+
+        zodb_conn = self.portal._p_jar
+
+        def commit():
+            conn = self.manager.connection
+            conn.commit(soft_commit=False, extract_after_commit=False)
+            zodb_conn.cacheGC()
+            self.log(
+                'Intermediate commit (%d items processed, last batch in %s)',
+                processed, lap.next())
+
+        cpi = checkpoint_iterator(commit, interval=commit_interval)
+        self.log('Syncing modified to solr for {} objects'.format(len(to_correct)))
+        for uid in to_correct:
+            catalog_item = catalog.unrestrictedSearchResults(UID=uid)[0]
+
+            handler = getMultiAdapter((catalog_item, self.manager), ISolrIndexHandler)
```

Comment:
i'm a bit confused as to why this works at all for a brain?

Reply:
There is a handler registered for `*`